File: core/src/main/java/com/toocol/ssh/core/mosh/core/crypto/Crypto.java
package com.toocol.ssh.core.mosh.core.crypto;

import com.toocol.ssh.utilities.execeptions.CryptoException;
import com.toocol.ssh.utilities.utils.MessageBox;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * crypto.cc
 *
 * @author ：JoeZane (joezane.cn@gmail.com)
 * @date: 2022/4/30 17:26
 * @version: 0.0.1
 */
@SuppressWarnings("all")
public final class Crypto {

    public static class Nonce {
        public static final int NONCE_LEN = 12;

        private final byte[] bytes = new byte[NONCE_LEN];

        public Nonce(long directionSeq) {
            System.arraycopy(ByteOrder.htoBe64(directionSeq), 0, this.bytes, 4, 8);
        }

        public Nonce(byte[] bytes, int len) {
            if (len != 8) {
                throw new CryptoException("Nonce representation must be 8 octets long.");
            }

            System.arraycopy(bytes, 0, this.bytes, 4, 8);
        }

        public long val() {
            byte[] longBytes = new byte[8];
            System.arraycopy(this.bytes, 4, longBytes, 0, 8);
            return ByteOrder.be64toh(longBytes);
        }

        public byte[] data() {
            return bytes;
        }

        public byte[] ccBytes() {
            byte[] cc = new byte[8];
            System.arraycopy(bytes, 4, cc, 0, 8);
            return cc;
        }
    }

    public static class Message {
        public Nonce nonce;
        public byte[] text;

        public Message() {
        }

        public Message(Nonce nonce, byte[] text) {
            this.nonce = nonce;
            this.text = text;
        }

        public Message resetData(Nonce nonce, byte[] text) {
            this.nonce = nonce;
            this.text = text;
            return this;
        }

        public short getTimestamp() {
            return ByteOrder.be16toh(new byte[]{text[0], text[1]});
        }

        public short getTimestampReply() {
            return ByteOrder.be16toh(new byte[]{text[2], text[3]});
        }
    }

    public static class Base64Key {
        private static final int KEY_LEN = 16;
        private final byte[] key;

        public Base64Key(String printableKey) {
            if (printableKey.length() != 22) {
                throw new CryptoException("Key must be 22 letters long.");
            }

            String base64 = printableKey + "==";
            key = Base64.getDecoder().decode(base64);
            if (key.length != KEY_LEN) {
                throw new CryptoException("Key must represent 16 octets.");
            }

            if (!printableKey.equals(printableKey())) {
                throw new CryptoException("Base64 key was not encoded 128-bit key.");
            }
        }

        public String printableKey() {
            byte[] base64 = Base64.getEncoder().encode(key);

            if (base64[23] != '=' || base64[22] != '=') {
                throw new CryptoException("Unexpected output from base64_encode: " + new String(base64));
            }

            return new String(base64, 0, 22);
        }
    }

    public static class AlignedBuffer {
        private int len;
        private byte[] data;

        public AlignedBuffer(int len) {
            this.len = len;
            this.data = new byte[len];
        }
    }

    public static class Session {
        public static final int RECEIVE_MTU = 2048;
        public static final int ADDED_BYTES = 16;

        AeOcb.AeCtx ctx;
        long blocksEncrypted;

        AlignedBuffer plaintextBuffer;
        AlignedBuffer ciphertextBuffer;
        AlignedBuffer nonceBuffer;

        public Session(Base64Key key) {
            this.blocksEncrypted = 0;
            this.plaintextBuffer = new AlignedBuffer(RECEIVE_MTU);
            this.ciphertextBuffer = new AlignedBuffer(RECEIVE_MTU);
            this.nonceBuffer = new AlignedBuffer(Nonce.NONCE_LEN);
            this.ctx = new AeOcb.AeCtx();

            if (AeOcb.AE_SUCCESS != AeOcb.aeInit(this.ctx, key.key, 16, 12, 16)) {
                throw new CryptoException("Could not initialize AES-OCB context.");
            }
        }

        public byte[] encrypt(Message plainText) {
            int ptLen = plainText.text.length;
            int ciphertextLen = ptLen + 16;

            assert ciphertextLen * 2 <= ciphertextBuffer.len;
            assert ptLen * 2 <= plaintextBuffer.len;

            System.arraycopy(plainText.text, 0, plaintextBuffer.data, 0, plainText.text.length);
            System.arraycopy(plainText.nonce.data(), 0, nonceBuffer.data, 0, Nonce.NONCE_LEN);

            if (ciphertextLen != AeOcb.aeEncrypt(
                    ctx,
                    nonceBuffer.data,
                    plaintextBuffer.data,
                    ptLen,
                    null,
                    0,
                    ciphertextBuffer.data,
                    null,
                    1
            )) {
                throw new CryptoException("aeEncrypt() returned error.");
            }

            blocksEncrypted += ptLen >> 4;
            if ((ptLen & 0xF) > 0) {
                blocksEncrypted++;
            }

            if ((blocksEncrypted >> 47) > 0) {
                throw new CryptoException("Encrypted 2^47 blocks.");
            }

            String text = new String(
                    ciphertextBuffer.data,
                    0,
                    ciphertextLen,
                    StandardCharsets.UTF_8
            );

            byte[] bytes = new byte[8 + ciphertextLen];
            System.arraycopy(plainText.nonce.ccBytes(), 0, bytes, 0, 8);
            System.arraycopy(ciphertextBuffer.data, 0, bytes, 8, ciphertextLen);

            return bytes;
        }

        public Message decrypt(byte[] str, int len, Message message) {
            if (len < 24) {
                throw new CryptoException("Ciphertext must contain nonce and tag.");
            }

            int bodyLen = len - 8;
            int ptLen = bodyLen - 16;

            if (ptLen < 0) {
                MessageBox.setExitMessage("Mosh error, invalid message length.");
                System.exit(-1);
            }

            assert bodyLen <= ciphertextBuffer.len;
            assert ptLen <= plaintextBuffer.len;

            Nonce nonce = new Nonce(str, 8);
            System.arraycopy(str, 8, ciphertextBuffer.data, 0, bodyLen);
            System.arraycopy(nonce.data(), 0, nonceBuffer.data, 0, Nonce.NONCE_LEN);

            if (ptLen != AeOcb.aeDecrypt(ctx, /* ctx */
                    nonceBuffer.data,      /* nonce */
                    ciphertextBuffer.data, /* ct */
                    bodyLen,               /* ct_len */
                    null,                  /* ad */
                    0,                     /* ad_len */
                    plaintextBuffer.data,  /* pt */
                    null,                  /* tag */
                    1)) {                  /* final */
                throw new CryptoException("Packet failed integrity check.");
            }

            byte[] text = new byte[ptLen];
            System.arraycopy(plaintextBuffer.data, 0, text, 0, ptLen);
            return message.resetData(nonce, text);
        }
    }

    private static long counter = 0;

    public synchronized static long unique() {
        return counter++;
    }

}

File: mosh/src/crypto/crypto.rs
#![allow(dead_code)]
use std::sync::atomic::AtomicU64;
use libs::ByteOrder;

use super::{AeCtx, AeOcb, AE_SUCCESS};
pub const KEY_LEN: usize = 16;
pub const NONCE_LEN: usize = 12;
pub const RECEIVE_MTU: usize = 2048;

///////////////// Crypto
static COUNTER: AtomicU64 = AtomicU64::new(0);
pub struct Crypto;
impl Crypto {
    pub fn unique() -> u64 {
        COUNTER.fetch_add(1, std::sync::atomic::Ordering::SeqCst)
    }
}

///////////////// Nonce
#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub struct Nonce {
    bytes: [u8; NONCE_LEN],
}
impl Nonce {
    pub fn from_seq(direction_seq: u64) -> Self {
        let be64 = ByteOrder::htobe64(direction_seq);
        let mut nonce = Nonce {
            bytes: [0; NONCE_LEN],
        };

        nonce.bytes[4..NONCE_LEN].copy_from_slice(&be64);

        nonce
    }

    pub fn from_bytes(bytes: &[u8], len: usize) -> Self {
        if len != 8 {
            panic!("Nonce representation must be 8 octets long.")
        }
        let mut nonce = Nonce {
            bytes: [0; NONCE_LEN],
        };

        nonce.bytes[4..NONCE_LEN].copy_from_slice(&bytes[0..8]);

        nonce
    }

    pub fn val(&self) -> u64 {
        let mut long_bytes = [0u8; 8];
        long_bytes.copy_from_slice(&self.bytes[4..NONCE_LEN]);
        ByteOrder::be64toh(long_bytes)
    }

    pub fn data(&self) -> &[u8] {
        &self.bytes
    }

    pub fn cc_bytes(&self) -> [u8; 8] {
        let mut cc = [0u8; 8];
        cc.copy_from_slice(&self.bytes[4..NONCE_LEN]);
        cc
    }
}

///////////////// Message
#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Message {
    pub nonce: Nonce,
    pub text: Vec<u8>,
}
impl Message {
    pub fn new(nonce: Nonce, text: Vec<u8>) -> Self {
        Message { nonce, text }
    }

    pub fn get_timestamp(&self) -> u16 {
        ByteOrder::be16toh([self.text[0], self.text[1]])
    }

    pub fn get_timestamp_reply(&self) -> u16 {
        ByteOrder::be16toh([self.text[2], self.text[3]])
    }
}

///////////////// Base64Key
pub struct Base64Key {
    key: [u8; KEY_LEN],
}
impl Base64Key {
    pub fn new(printable_key: String) -> Self {
        let mut printable_key = printable_key;
        if printable_key.len() != 22 {
            panic!("Key must be 22 letters long.")
        };

        printable_key.push_str("==");
        let key = base64::decode(printable_key.as_bytes()).expect("Decode Base64Key failed.");
        if key.len() != KEY_LEN {
            panic!("Key must represent 16 octets.")
        };

        let mut base_64_key = Base64Key { key: [0; KEY_LEN] };
        for i in 0..KEY_LEN {
            base_64_key.key[i] = key[i]
        }

        if !printable_key.eq(&base_64_key.printable_key()) {
            panic!("Base64 key was not encoded 128-bit key.")
        }

        base_64_key
    }

    pub fn key(&self) -> [u8; KEY_LEN] {
        self.key
    }

    pub fn printable_key(&self) -> String {
        let base64_string = base64::encode(&self.key);
        let base64 = base64_string.as_bytes();

        if base64[22] != '=' as u8 || base64[23] != '=' as u8 {
            panic!("Unexpected output from base64_encode: {}", base64_string)
        }
        base64_string
    }
}

///////////////// AlignedBuffer
pub struct AlignedBuffer<const D: usize> {
    pub len: usize,
    pub data: [u8; D],
}
impl<const D: usize> AlignedBuffer<D> {
    pub fn new(data: [u8; D]) -> Self {
        AlignedBuffer {
            len: data.len(),
            data,
        }
    }

    pub fn len(&self) -> usize {
        self.len
    }
}

///////////////// Session
pub struct Session {
    ctx: AeCtx,
    blocks_encrypted: u64,

    plain_text_buffer: AlignedBuffer<RECEIVE_MTU>,
    cipher_text_buffer: AlignedBuffer<RECEIVE_MTU>,
    nonce_buffer: AlignedBuffer<NONCE_LEN>,
}
impl Session {
    pub const ADDED_BYTES: usize = 16;

    pub fn new(key: Base64Key) -> Self {
        let mut session = Session {
            ctx: AeCtx::new(),
            blocks_encrypted: 0,
            plain_text_buffer: AlignedBuffer::new([0u8; RECEIVE_MTU]),
            cipher_text_buffer: AlignedBuffer::new([0u8; RECEIVE_MTU]),
            nonce_buffer: AlignedBuffer::new([0u8; NONCE_LEN]),
        };

        if AE_SUCCESS != AeOcb::ae_init(&mut session.ctx, key.key(), 16, 12, 16) {
            panic!("Could not initialize AES-OCB context.")
        }
        session
    }

    pub fn encrypt(&mut self, plain_text: Message) -> Vec<u8> {
        let pt_len = plain_text.text.len();
        let cipher_text_len = pt_len + 16;

        assert!(cipher_text_len * 2 <= self.cipher_text_buffer.len());
        assert!(pt_len * 2 <= self.plain_text_buffer.len());

        self.plain_text_buffer.data[0..plain_text.text.len()]
            .copy_from_slice(&plain_text.text[0..plain_text.text.len()]);
        self.nonce_buffer.data[0..NONCE_LEN].copy_from_slice(plain_text.nonce.data());

        if cipher_text_len
            != AeOcb::ae_encrypt(
                &mut self.ctx,
                &mut self.nonce_buffer.data,
                &mut self.plain_text_buffer.data,
                pt_len,
                None,
                0usize,
                &mut self.cipher_text_buffer.data,
                None,
                1,
            ) as usize
        {
            panic!("aeEncrypt() returned error.")
        }

        self.blocks_encrypted += pt_len as u64 >> 4;
        if (pt_len & 0xF) > 0 {
            self.blocks_encrypted += 1;
        }

        if (self.blocks_encrypted >> 47) > 0 {
            panic!("Encrypted 2^47 blocks.")
        }

        let mut bytes: Vec<u8> = vec![0; cipher_text_len + 8];
        let cc_bytes = plain_text.nonce.cc_bytes();
        bytes[0..8].copy_from_slice(&cc_bytes[0..8]);
        bytes[8..8 + cipher_text_len]
            .copy_from_slice(&self.cipher_text_buffer.data[0..cipher_text_len]);

        bytes
    }

    pub fn decrypt(&mut self, str: &[u8], len: usize) -> Message {
        if len < 24 {
            panic!("Ciphertext must contain nonce and tag.")
        }

        let body_len = len - 8;
        if body_len < 16 {
            panic!("Mosh error, invalid message length.")
        }
        let pt_len = body_len - 16;

        assert!(body_len <= self.cipher_text_buffer.len());
        assert!(pt_len <= self.plain_text_buffer.len());

        let nonce = Nonce::from_bytes(str, 8);
        self.cipher_text_buffer.data[0..body_len].copy_from_slice(&str[8..8 + body_len]);
        self.nonce_buffer.data[0..NONCE_LEN].copy_from_slice(&nonce.data()[0..NONCE_LEN]);

        if pt_len
            != AeOcb::ae_decrypt(
                &mut self.ctx,
                &self.nonce_buffer.data,
                &self.cipher_text_buffer.data,
                body_len as i32,
                None,
                0usize,
                &mut self.plain_text_buffer.data,
                None,
                1,
            ) as usize
        {
            panic!("Packet failed integrity check.")
        }

        let mut text: Vec<u8> = vec![0; pt_len];
        text[0..pt_len].copy_from_slice(&self.plain_text_buffer.data[0..pt_len]);
        Message::new(nonce, text)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_base_64_key() {
        let _key = Base64Key::new("zr0jtuYVKJnfJHP/XOOsbQ".to_string());
    }

    #[test]
    fn test_encrypt_decrypt() {
        let key = Base64Key::new("zr0jtuYVKJnfJHP/XOOsbQ".to_string());
        let mut session = Session::new(key);

        let plain_text = "Rust is a multi-paradigm, general-purpose programming language. Rust emphasizes performance, type safety, and concurrency.[11][12][13] 
            Rust enforces memory safety—that is, that all references point to valid memory—without requiring the use of a garbage collector or reference counting present in other memory-safe languages.[13][14]";
        let nonce = Nonce::from_seq(10);

        let bytes = plain_text.as_bytes();
        let mut text: Vec<u8> = vec![0; bytes.len()];
        text[0..bytes.len()].copy_from_slice(bytes);

        let en_message = Message::new(nonce, text);
        let encrypted = session.encrypt(en_message.clone());
        let de_message = session.decrypt(&encrypted, encrypted.len());
        assert_eq!(en_message, de_message);
    }
}
